[MRG] Landmark ISOMAP

lucasdavid opened the pull request:
First implementation for Landmark Isomap (L-Isomap). References issue #5935.

Commit:
upgrade Isomap to perform L-Isomap as well

lucasdavid commented:
I'm getting an error on something that I haven't even touched:
```
================================
Gradient Boosting regularization
================================

[...]

Subsampling witTraceback (most recent call last):
  File "/home/ubuntu/scikit-learn/doc/sphinxext/gen_rst.py", line 864, in generate_file_rst
    execfile(os.path.basename(src_file), my_globals)
  File "plot_partial_dependence.py", line 61, in <module>
    cal_housing = fetch_california_housing()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sklearn/datasets/california_housing.py", line 93, in fetch_california_housing
    fhandle = urlopen(DATA_URL)
  File "/usr/lib/python2.7/urllib2.py", line 127, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 404, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 422, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1214, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/usr/lib/python2.7/urllib2.py", line 1184, in do_open
    raise URLError(err)
URLError: <urlopen error [Errno -2] Name or service not known>
hout shrinkage usually does poorly.
Another strategy to reduce the variance is by subsampling the features
analogous to the random splits in Random Forests
(via the ``max_features`` parameter).
```

jakevdp commented:
One overall comment: the ``shortest_path`` code is effectively an out-dated back-port of what exists in the ``scipy.sparse.csgraph`` package. The ``only_vertices`` keyword you added here is equivalent to the ``indices`` keyword that's been implemented there for several releases. If you need this functionality, I'd recommend making use of the code that's already within scipy.

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -196,16 +195,192 @@ def transform(self, X):
-            G_X[i] = np.min((self.dist_matrix_[indices[i]]
-                             + distances[i][:, None]), 0)
+            G_X[i] = np.min(self.dist_matrix_[:, indices[i, :]] + distances[i],
+                            axis=1)
```
Curious why you modified the ``Isomap`` code here. It looks like you're grabbing columns of the distance matrix rather than rows – is there a particular reason for that change?

lucasdavid replied:
Yes. L-Isomap only keeps a `dist_matrix_` of size n_landmarks x N, so not all distances are defined in the rows. If I hadn't modified Isomap, I would have to implement a new transform method.

jakevdp replied:
Got it – that makes sense!

jakevdp replied:
Another option would be to transpose the distance matrix in L-Isomap... there's a small (arguably negligible) chance that this change could break someone's code if they are extending Isomap. If the transpose solution is easy, it may be worth the switch.

lucasdavid replied:
I see. Yes, I think this can be made pretty easily.

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -196,16 +195,192 @@ def transform(self, X):
+                # should be strictly superior to the number of dimensions
+                # of the reduced data set added by some safety margin.
+                n_landmarks = self.n_components + 1 \
+                              + min(self.n_components // 2, 100)
```
Style thing: I'd go for line wrapping in parentheses rather than with a slash-escape.

jakevdp commented on sklearn/utils/estimator_checks.py:
```diff
@@ -555,8 +556,6 @@ def _check_transformer(name, Transformer, X, y):
     set_testing_parameters(transformer)
 
-    # fit
-
```
Probably revert changes here, as they are unrelated to the PR.

jakevdp commented:
Overall looks really good and clean! It's nice that so much of the standard Isomap can be re-used in the inherited class. A couple general notes:

- As I mentioned, ``scipy.sparse.csgraph`` includes additional functionality similar to what you added here (and has since early 2013). Rather than letting this backport diverge, I think it would be better to update the backport. Someday when the scipy dependency is mature enough, we should be aiming to drop this code from scikit-learn

- There is still some code duplication between ``Isomap`` and ``LandmarkIsomap``. It might be worth refactoring ``Isomap`` slightly so that these differences are in a separate utility routine, so that it's even clearer on reading the code where the two algorithms diverge. Or perhaps, as it's such a small change overall, we could simply add a "landmark" keyword to ``Isomap`` itself. This is the approach taken for this type of variant in, e.g. ``LocallyLinearEmbedding``. I could go either way though.

- Also there is a fair amount of code duplication between the Isomap/LandmarkIsomap tests. I'd be in favor of moving to a single common test suite for both, along with any additional tests that are needed for landmark.

- Once these API decisions are made, it would be good to add some mention of this code in the narrative documentation on isomap, as well as linking to your excellent plot demo of the differences between them.

Nice work! Good to see this coming along.

lucasdavid commented:
Ok! Thank you so much for the suggestions, I will implement the changes as soon as possible!

jakevdp commented:
Actually, I'd like someone (maybe @GaelVaroquaux or @ogrisel) to weigh-in on the API question. Should we prefer ``LandmarkIsomap()`` or ``Isomap(method='landmark')`` here? I used the latter style in ``LocallyLinearEmbedding``, so there's precedent for that in the submodule, but I might lean toward the former now.

lucasdavid commented:
Using a single class seems nice, as they do share a lot of code. One thing I'm not too psyched about, though, is that it will create the possibility of semantically conflicting parameters: what happens when someone call `Isomap(method=None, n_landmarks=10)`, or the equivalent `Isomap(n_landmarks=10)`? Should it run L-Isomap or the original one? Or should it raise an warning or error saying that the `n_landmarks` parameter doesn't make sense because `method=None` means running the original algorithm? Or run should it simply run the original one and don't say anything? That strikes me as a little bit confusing.

jakevdp commented:
Good point – the new class is probably a cleaner option.

jakevdp commented:
It occurred to me that one way to do this "all in one" is to just add a ``landmarks=None`` parameter to ``Isomap``. In this case, the default would be to use all points as landmarks, which is equivalent to standard Isomap, if I'm not mistaken.

lucasdavid commented:
Okay, I didn't think that was a possibility. `Isomap(n_landmarks=10)` and `Isomap(landmarks=[0, 10, 4, ...])` seem great!
Yes, `n_landmarks==n_samples` makes L-Isomap equivalent to the original algorithm.

I'll get right on it, but first, just to be sure:
```py
n_landmarks=None   # (default): use the original Isomap.
n_landmarks='auto' # Compute number of landmarks to use.
n_landmarks=k      # Use k samples as landmarks.
landmarks=[...]    # Ignore above and use these elements as landmarks.
```

jakevdp commented:
> just to be sure
> ...

Yes, that's what I have in mind. I think it will end up being a very nice & clean API! Thanks for being willing to bounce around on this a bit :grinning:

Commit:
update graph module and its tests

lucasdavid commented:
No problem!
I've merged the algorithms, reverted `Isomap.transform()` changes and removed `graph_shortest_path` module, keeping only `graph.shortest_path` as wrapper for the scipy functions. It does seem a little bit cleaner now.

I'm having a weird problem with travis: I deleted the `test_landmark_isomap.py` file but for some reason it's still being executed. Nose doesn't point out any errors on my machine. I'll come back to this tomorrow.

jakevdp commented:
Sorry, I should have been more clear. We can't remove ``graph_shortest_path`` now, because our minimum SciPy dependency is currently 0.9 and the functionality wasn't added until 0.11 (though it may be a good time to bump the requirement: @GaelVaroquaux usually makes the call on those kinds of things). What we will need instead is to update the backport in scikit-learn with the additional functionality available in scipy 0.11. Hope that makes sense.

lucasdavid commented:
Oh I see. Sorry for the mistake. I'll roll back the changes.

Commit:
update landmarks selection strategies

lucasdavid commented:
Done. Everything is working fine now, but I'm not really happy with copying code like this.

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -45,19 +45,63 @@ class Isomap(BaseEstimator, TransformerMixin):
+
+    landmarks : array-like, optional (default = None)
+        The list of landmarks to use. If none was passed,
+        infers it from n_landmarks.
```
Rather than both ``n_landmarks`` and ``landmarks``, I think it would make sense to have a single parameter ``landmarks`` which is either ``None``, ``'auto'``, an integer (specifying number of landmarks), or an array-like (specifying the indices of the landmarks).

lucasdavid replied:
I thought about it last night and I agree. We still need `landmarks_` attribute, though, was a user might pass an integer and the actual array is still needed for future reference.

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -96,11 +149,51 @@ def __init__(self, n_neighbors=5, n_components=2, eigen_solver='auto',
         self.nbrs_ = NearestNeighbors(n_neighbors=n_neighbors,
                                       algorithm=neighbors_algorithm)
+        self.n_landmarks = n_landmarks
+        self.landmarks_ = landmarks
```
The ``init`` function should store each named parameter by name. This is so that the ``clone()`` functionality will work correctly.

jakevdp replied:
On that topic, I just added this fix to the issue that was already there: #6062

lucasdavid replied:
Done! :-)

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -96,11 +149,51 @@ def __init__(self, n_neighbors=5, n_components=2, eigen_solver='auto',
         self.nbrs_.fit(X)
-        self.training_data_ = self.nbrs_._fit_X
+        self.training_data_ = self.nbrs_._fit_X if landmarks is None else self.nbrs_._fit_X[landmarks]
+
```
I think we should leave the training data as the full ``X`` data. If you need an attribute storing the landmarks, create another one.

jakevdp replied:
Also, what happens if the ``landmarks`` has repeats in it? There is no check for this yet, but in this case ``training_data_`` will have repeated samples.

lucasdavid replied:
No problem. But just to be clear: if training data is kept entirely, we need to change the transform method.
```py
G_X = np.zeros((X.shape[0], self.training_data_.shape[0]))
# to something like ->
G_X = np.zeros((X.shape[0],
                self.training_data_.shape[0] if landmarks is None
                else landmarks.shape[0]))
```

jakevdp replied:
Yes – I understand that you'll have to use this array differently, but to me it makes more sense for ``training_data`` to store the training data, rather than the landmarks.

lucasdavid replied:
Well, that's truth, but I'm not sure if this should be checked. We do not verify if there are repeated samples in the original data set either.

jakevdp replied:
fair enough!

lucasdavid replied:
Yes, you're right about the `training_data_`.

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -109,13 +202,29 @@ def _fit_transform(self, X):
-        G = self.dist_matrix_ ** 2
-        G *= -0.5
+        self.dist_matrix_ = shortest_path(kng,
+                                          method='D',
```
should use ``self.path_method`` here.

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -109,13 +202,29 @@ def _fit_transform(self, X):
+        self.dist_matrix_ = shortest_path(kng,
+                                          method='D',
+                                          directed=False,
+                                          indices=landmarks).T
```
Only pass ``landmarks`` if necessary (as currently written, it's slightly slower for the default case)

jakevdp replied:
Wait, strike that – I see that ``landmarks=None`` by default. This is OK

jakevdp commented on sklearn/manifold/isomap.py:
```diff
@@ -109,13 +202,29 @@ def _fit_transform(self, X):
+                                          indices=landmarks).T
 
-        self.embedding_ = self.kernel_pca_.fit_transform(G)
+        self.dist_matrix_[np.where(np.isinf(self.dist_matrix_))] = 0
```
Why this step? Also, ``self.dist_matrix_[np.isinf(self.dist_matrix_)] = 0`` will be more efficient.

lucasdavid replied:
Without this, I'm getting this error in `sklearn/tests/test_commons.py`:
```shell
ERROR: sklearn.tests.test_common.test_non_meta_estimators('Isomap', <class 'sklearn.manifold.isomap.Isomap'>)
# ...
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

My guess was that scipy's functions didn't replace inf with zeros at the end of the methods, as sklearn's [version does](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/graph_shortest_path.pyx#L152). I didn't want to modify this in `shortest_path` because that's a complete logic change.

jakevdp commented on sklearn/utils/graph_shortest_path.pyx:
```diff
@@ -1,238 +1,1118 @@
     undirected graph.
 
+    .. versionadded:: 0.11.0
+
```
This versionadded refers to scipy; it should be taken out here.

Also perhaps add a note that this function is backported from scipy.

lucasdavid replied:
I did add a note at the [beginning](https://github.com/scikit-learn/scikit-learn/pull/5969/files#diff-3c1994dba253eb42a1ee64949bf2b9ccR9) of the file. Should I move it to the functions docs?

jakevdp commented on sklearn/utils/graph_shortest_path.pyx:
```diff
@@ -1,238 +1,1118 @@
+
+    Compute the shortest path lengths using the Floyd-Warshall algorithm
+
+    .. versionadded:: 0.11.0
```
Same here: this refers to SciPy.

jakevdp commented:
One thing here: this would amount to a silent removal of ``sklearn.utils.graph_shortest_path``, because it's now renamed to ``shortest_path``. This is more a question for one of the other core devs (ping @GaelVaroquaux, @ogrisel, @amueller): should we worry about renaming a backport within ``utils`` and/or changing its API?

lucasdavid commented:
I renamed it to match scipy's because, according to pycharm's refactoring tool, the only references to `graph_shortest_path` are in the following places (branch master):
```
graph_shortest_path(dist_matrix, directed=True, method='auto')
scikit-learn
    .gitattributes
        /sklearn/utils/graph_shortest_path.c -diff
    doc/developers
        utilities.rst
              iteration of Dijkstra's algorithm from ``graph_shortest_path``.
            - :func:`graph_shortest_path.graph_shortest_path`:
            - :func:`graph_shortest_path.graph_shortest_path`:
sklearn/manifold
    isomap.py
        Isomap
            _fit_transform
                self.dist_matrix_ = graph_shortest_path(kng,
sklearn/utils/tests
    test_shortest_path.py
sklearn/manifold
    isomap.py
        from ..utils.graph import graph_shortest_path
    sklearn/utils
        graph.py
            from .graph_shortest_path import graph_shortest_path
    sklearn/utils/tests
        test_shortest_path.py
            from sklearn.utils.graph import (graph_shortest_path,
```
As you can see, all these guys are the `graph.py`, docs, tests, or Isomap itself.

jakevdp commented:
> the only references to graph_shortest_path...

I see – my only concern with renaming it is if someone is using this utility externally. I recall there being a recent discussion regarding whether functionality in ``utils`` should be considered stable from release to release or subject to change without notice. I think it's the latter, but I don't remember what resolution the team came to!

lucasdavid commented:
Q: Is there a reason for `Isomap._fit_transform` method to exist? It is basically `fit_transform` method without the `return self` statement. Can't we just merge both of them?

---

I've implemented the landmarks selection with using a min-max strategy for, described in de Silva and Tenenbaum's [paper](http://www.cmap.polytechnique.fr/~peyre/cours/x2005signal/dimreduc_landmarks.pdf), which helps selecting well-distributed landmarks that results in better embeddings.
In the figure bellow, the second column contains the embeddings found with `landmark_selection_method='random'`, whereas the third column has the embeddings found with `landmark_selection_method='min-max'`.
![fig2](https://cloud.githubusercontent.com/assets/1595030/11907035/f219f5aa-a5b7-11e5-9d1f-98a2cb242081.png)

Also, `'min-max'` and `'random'` options impact little in the algorithm's overall performance:
```
Comparison 0: 1000000 samples
34.3275 sec: #0: {'n_neighbors': 10, 'landmark_selection_method': 'random', 'landmarks': 'auto', 'n_components': 2}
37.6842 sec: #1: {'n_neighbors': 10, 'landmark_selection_method': 'min-max', 'landmarks': 'auto', 'n_components': 2}

Comparison 1: 6000000 samples
288.049 sec: #0: {'n_neighbors': 10, 'landmark_selection_method': 'random', 'landmarks': 'auto', 'n_components': 2}
279.265 sec: #1: {'n_neighbors': 10, 'landmark_selection_method': 'min-max', 'landmarks': 'auto', 'n_components': 2}

Comparison 2: 11000000 samples
518.753 sec: #0: {'n_neighbors': 10, 'landmark_selection_method': 'random', 'landmarks': 'auto', 'n_components': 2}
536.051 sec: #1: {'n_neighbors': 10, 'landmark_selection_method': 'min-max', 'landmarks': 'auto', 'n_components': 2}

Comparison 3: 16000000 samples
744.108 sec: #0: {'n_neighbors': 10, 'landmark_selection_method': 'random', 'landmarks': 'auto', 'n_components': 2}
725.72 sec: #1: {'n_neighbors': 10, 'landmark_selection_method': 'min-max', 'landmarks': 'auto', 'n_components': 2}

Comparison 4: 21000000 samples
MemoryError.
```

I can't see any other way to improve this algorithm right now, but let me know if there is or if further changes should be made. Otherwise, I guess I'll start changing the docs.

Commit:
add L-Isomap docs and some refactoring

ogrisel commented:
Please do not merge the master branch into a PR, that makes it harder to squash intermediate commit later...

Instead it would be better if you could under the merge commit, squash all the non interesting intermediate commits and rebase on master to get travis to run the tests.

ogrisel commented:
However if you are not comfortable with git, don't do and a maintainer will do that prior to merging the PR instead.

lucasdavid commented:
No problem, I can do that.

lucasdavid commented:
@ogrisel: done! I managed to revert the merging with master and squash the commits.

Unrelated: I'm getting the error *Your build has exceeded the memory limit of 4G on 1 container [...]* from circleci and the following in the output:
```
./continuous_integration/circle/build_doc.sh died unexpectedly
Done. Reconstruction errAction failed: ./continuous_integration/circle/build_doc.sh
```

@ogrisel and @jakevdp: any thoughts on what is causing this? I haven't changed `build_doc.sh` at all, but I did add a reference to my `plot_compare_isomaps.py` in `doc/modules/manifold.rst` (I wonder if this is the problem):
```
* See :ref:`example_manifold_plot_compare_isomaps.py` for a performance
comparison between the original Isomap and L-Isomap.
```

jakevdp commented on examples/manifold/plot_compare_isomaps.py:
```diff
@@ -0,0 +1,83 @@
+        n_samples,
+        (manifold.Isomap, isomap_params),
+        (manifold.Isomap, l_isomap_params),
+    ) for n_samples in range(1000, 16001, 5000)]
```
n_samples=16000 might lead to a large memory use, if this code is run by the circleCI build – it would create two 16000x16000 matrices – one in each isomap approach.

jakevdp replied:
Maybe you meant this to be 1600?

lucasdavid replied:
The idea here was to show how the difference between both algorithms' performances increases as the number of samples increases, and how L-Isomap can handle this high number of samples whereas Isomap cannot. To avoid memory collapse the fitting happens inside a [try/raise](https://github.com/scikit-learn/scikit-learn/pull/5969/files#diff-a01c99fa301b88ee5a62d79240cfe3e8R78) clause:
```py
try:
    t0, t1 = time(), None
    ...
except MemoryError as e:
    print('Failed.' % j)
```

This should be protecting the build against memory errors, shouldn't it? Anyhow, I'll reduce and see if it works.

jakevdp replied:
I think the memory error is happening in the system, outside of Python, so a Python try/except is not going to catch it.

lucasdavid replied:
You were right. CircleCI build is now passing as well!

lucasdavid commented:
@ogrisel I finished all the changes and it's been a while now. Do you think we could have this merged? Thanks!

amueller commented:
Can you please fix the flake8 errors? I'll try to catch up with the reviewing but I can't promise anything.

amueller commented:
@jakevdp do you want to review?

vals commented:
Is there any progress on this? I think it would be a great addition...

lucasdavid commented:
@vals if you are working with nonlinear embedding you might want to experiment with multi-layered encoders/convolutional encoders with non-linear activation functions. Those yielded pretty interesting results in my few experiments. Furthermore, they are incremental like any network, which deals with the problem L-ISOMAP tries to tackle.

noshita commented:
I'm trying to figure out why this PR is closed. Do anyone continue to work on this?

jnothman commented:
From the looks of things, @lucasdavid lost interest in contributing this, perhaps because convolutional encoders worked well for him instead. But I think it mostly just needs some review and a contributor to take over.

lucasdavid commented:
@jnothman I just got bored of rebasing and no one reviewing. I don't blame you guys, obviously. I understand this is not a hot feature and there are other section that need more attention.

About contributing to this: I think it would be less painful if another pull-request were made, updating scipy's graph backdoor. This will reduce the modified files count of this PR from 20 to 4 and make it much easier to review.

Antonio-Leitao commented:
Landmark MDS and ISOMAP are very powerful implementations that would surely benefit the community. I'm considering restarting this project if there is interest.

vals commented:
Curiously, though it took me a couple of years, I also ended up moving to autoencoder methods.
